packages/service-worker/worker/src/data.ts

```
@@ -624,4 +622,38 @@ export class DataGroup {
       });
     }
   }
+
+  /**
+   * Detect if storage is full or approaching capacity.
+   * Returns true if storage is at or near capacity.
+   */
+  private async detectStorageFull() {
+    try {
+      const estimate = await navigator.storage.estimate();
+      const {quota, usage} = estimate;
+
+      // Handle cases where quota or usage might be undefined
+      if (typeof quota !== 'number' || typeof usage !== 'number') {
+        return;
+      }
+
+      // Consider storage "full" if usage is >= 95% of quota
+      // This provides a safety buffer before actual storage exhaustion
+      const usagePercentage = (usage / quota) * 100;
+      const isStorageFull = usagePercentage >= 95;
+
+      if (isStorageFull) {
+        this.debugHandler.log(
+          'Storage is full or nearly full',
+          `DataGroup(${this.config.name}@${this.config.version}).detectStorageFull()`,
+        );
+      }
+    } catch (err) {
+      // Storage API call failed
+      this.debugHandler.log(
```

If the storage API is throwing, I can imagine it will always throw. Might it be better to just ignore the error or at least ensure this isn't logged more than once?

---

I think ignoring the error is reasonable due to baseline support. This is also only a debug/helpful check, not something that's required for things to work so the error doesn't _need_ to be reported

---

@atscott  updated it